python/ray/tune/trial_runner.py

```diff
@@ -69,6 +71,119 @@ def _from_cloudpickle(self, obj):
         return cloudpickle.loads(hex_to_binary(obj["value"]))
 
 
+class Callback:
```

Comment:
Should we avoid calling back into the trialrunner? the trial runner seems to be an internal component (i.e., its methods are not publicly documented).

Reply:
The main question then is if there are any properties that users might want to deal with. E.g. should we then pass the list of `trials`, or other pieces of the state that people might want to interact with.

Reply:
Trials are good. Can we do it in a forward compat way?

Reply:
After our offline discussion I thought about the possibilities that we have and decided to go with a keyword argument parameter. In my opinion we retain maximum flexibility with this approach while ensuring forwards compatibility:

- Arguments can easily be added without breaking existing code
- Arguments can be deprecated and raise a warning if they are used easily
- No state in callbacks is retained, child classes can define their own constructor without having to call `super()`
- Clear documentation of keyword arguments is easier than documentation of class attributes

I updated the PR with that.